phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java

```diff
@@ -885,292 +867,234 @@ private static boolean clipKeyRangeBytes(RowKeySchema schema, int fieldIndex, in
+                if (index < ignoredRegionList.size() && ignoredRegionList.get(index) == i) {
+                    index++;
+                    continue;
+                }
+
+                HRegionInfo regionInfo = regionLocations.get(i).getRegionInfo();
+
+                for (KeyRange queryKeyRange : queryKeyRanges) {
+                    byte[] bytesToPrepend = regionInfo.getStartKey();
+                    int length = bytesToPrepend.length;
+                    if (length == 0) {
+                        length = regionInfo.getEndKey().length;
+                        if (length > 0) {
+                            bytesToPrepend = new byte[length];
+                        }
+                    }
+
+                    KeyRange newQueryRowKeyRange = queryKeyRange.prependRange(
+                            bytesToPrepend, 0, bytesToPrepend.length);
+                    newQueryRowKeyRanges.add(newQueryRowKeyRange);
+                }
+            }
+
+            if (newQueryRowKeyRanges.size() == 0) {
+                newQueryRowKeyRanges.add(KeyRange.EMPTY_RANGE);
+            }
+
+            queryKeyRanges = newQueryRowKeyRanges;
+        }
+
+        return queryKeyRanges;
+    }
+
     /**
      * Compute the list of parallel scans to run for a given query. The inner scans
      * may be concatenated together directly, while the other ones may need to be
-     * merge sorted, depending on the query.
-     * Also computes an estimated bytes scanned, rows scanned, and last update time
-     * of statistics. To compute correctly, we need to handle a couple of edge cases:
-     * 1) if a guidepost is equal to the start key of the scan.
-     * 2) If a guidepost is equal to the end region key.
-     * In both cases, we set a flag (delayAddingEst) which indicates that the previous
-     * gp should be use in our stats calculation. The normal case is that a gp is
-     * encountered which is in the scan range in which case it is simply added to
-     * our calculation.
-     * For the last update time, we use the min timestamp of the gp that are in
-     * range of the scans that will be issued. If we find no gp in the range, we use
-     * the gp in the first or last region of the scan. If we encounter a region with
-     * no gp, then we return a null value as an indication that we don't know with
-     * certainty when the stats were updated last. This handles the case of a split
-     * occurring for a large ingest with stats never having been calculated for the
-     * new region.
+     * merge sorted, depending on the query. Also computes an estimated bytes scanned,
+     * rows scanned, and last update time of statistics.
+     *
      * @return list of parallel scans to run for a given query.
      * @throws SQLException
      */
     private List<List<Scan>> getParallelScans(byte[] startKey, byte[] stopKey) throws SQLException {
+        ScanRanges scanRanges = context.getScanRanges();
```

Comment:
why we change the order here?

Reply:
The whole function was actually rewritten.